// angular-auth-oidc-client/esm2022/lib/extractors/jwk.extractor.mjs
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
class JwkExtractor {
    extractJwk(keys, spec, throwOnEmpty = true) {
        if (0 === keys.length) {
            throw JwkExtractorInvalidArgumentError;
        }
        const foundKeys = keys
            .filter((k) => (spec?.kid ? k['kid'] === spec.kid : true))
            .filter((k) => (spec?.use ? k['use'] === spec.use : true))
            .filter((k) => (spec?.kty ? k['kty'] === spec.kty : true));
        if (foundKeys.length === 0 && throwOnEmpty) {
            throw JwkExtractorNoMatchingKeysError;
        }
        if (foundKeys.length > 1 && (null === spec || undefined === spec)) {
            throw JwkExtractorSeveralMatchingKeysError;
        }
        return foundKeys;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.0.2", ngImport: i0, type: JwkExtractor, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "16.0.2", ngImport: i0, type: JwkExtractor, providedIn: 'root' }); }
}
export { JwkExtractor };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.0.2", ngImport: i0, type: JwkExtractor, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }] });
function buildErrorName(name) {
    return JwkExtractor.name + ': ' + name;
}
export const JwkExtractorInvalidArgumentError = {
    name: buildErrorName('InvalidArgumentError'),
    message: 'Array of keys was empty. Unable to extract',
};
export const JwkExtractorNoMatchingKeysError = {
    name: buildErrorName('NoMatchingKeysError'),
    message: 'No key found matching the spec',
};
export const JwkExtractorSeveralMatchingKeysError = {
    name: buildErrorName('SeveralMatchingKeysError'),
    message: 'More than one key found. Please use spec to filter',
};
