// angular-auth-oidc-client/esm2022/lib/storage/default-sessionstorage.service.mjs
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
class DefaultSessionStorageService {
    read(key) {
        return sessionStorage.getItem(key);
    }
    write(key, value) {
        sessionStorage.setItem(key, value);
    }
    remove(key) {
        sessionStorage.removeItem(key);
    }
    clear() {
        sessionStorage.clear();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.0.2", ngImport: i0, type: DefaultSessionStorageService, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "16.0.2", ngImport: i0, type: DefaultSessionStorageService, providedIn: 'root' }); }
}
export { DefaultSessionStorageService };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.0.2", ngImport: i0, type: DefaultSessionStorageService, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }] });
