// angular-auth-oidc-client/esm2022/lib/storage/default-localstorage.service.mjs
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
class DefaultLocalStorageService {
    read(key) {
        return localStorage.getItem(key);
    }
    write(key, value) {
        localStorage.setItem(key, value);
    }
    remove(key) {
        localStorage.removeItem(key);
    }
    clear() {
        localStorage.clear();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.0.2", ngImport: i0, type: DefaultLocalStorageService, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "16.0.2", ngImport: i0, type: DefaultLocalStorageService, providedIn: 'root' }); }
}
export { DefaultLocalStorageService };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.0.2", ngImport: i0, type: DefaultLocalStorageService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }] });
