// angular-auth-oidc-client/schematics/ng-add/schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FlowType = void 0;
var FlowType;
(function (FlowType) {
    FlowType["OidcCodeFlowPkceAzureAdUsingRefreshTokens"] = "OIDC Code Flow PKCE Azure AD using refresh tokens";
    FlowType["OidcCodeFlowPkceAzureAdUsingIframeSilentRenew"] = "OIDC Code Flow PKCE Azure AD using iframe silent renew";
    FlowType["OidcCodeFlowPkceUsingRefreshTokens"] = "OIDC Code Flow PKCE using refresh tokens";
    FlowType["OAuthPushAuthorizationRequestsUsingRefreshTokens"] = "OAuth Push authorization requests using refresh tokens";
    FlowType["OidcCodeFlowPkceUsingIframeSilentRenew"] = "OIDC Code Flow PKCE using iframe silent renew";
    FlowType["OidcCodeFlowPkceUsingIframeSilentRenewGettingConfigFromHttp"] = "OIDC Code Flow PKCE using iframe silent renew getting config from http";
    FlowType["OIDCCodeFlowPkce"] = "OIDC Code Flow PKCE (no renew)";
    FlowType["Auth0"] = "Auth0";
    FlowType["DefaultConfig"] = "Default config";
})(FlowType = exports.FlowType || (exports.FlowType = {}));
