// angular-auth-oidc-client/esm2022/lib/utils/flowHelper/flow-helper.service.mjs
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
class FlowHelper {
    isCurrentFlowCodeFlow(configuration) {
        return this.currentFlowIs('code', configuration);
    }
    isCurrentFlowAnyImplicitFlow(configuration) {
        return (this.isCurrentFlowImplicitFlowWithAccessToken(configuration) ||
            this.isCurrentFlowImplicitFlowWithoutAccessToken(configuration));
    }
    isCurrentFlowCodeFlowWithRefreshTokens(configuration) {
        const { useRefreshToken } = configuration;
        return this.isCurrentFlowCodeFlow(configuration) && useRefreshToken;
    }
    isCurrentFlowImplicitFlowWithAccessToken(configuration) {
        return this.currentFlowIs('id_token token', configuration);
    }
    currentFlowIs(flowTypes, configuration) {
        const { responseType } = configuration;
        if (Array.isArray(flowTypes)) {
            return flowTypes.some((x) => responseType === x);
        }
        return responseType === flowTypes;
    }
    isCurrentFlowImplicitFlowWithoutAccessToken(configuration) {
        return this.currentFlowIs('id_token', configuration);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.0.2", ngImport: i0, type: FlowHelper, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "16.0.2", ngImport: i0, type: FlowHelper, providedIn: 'root' }); }
}
export { FlowHelper };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.0.2", ngImport: i0, type: FlowHelper, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }] });
